### PR TITLE
🐛 Use patch helper to update BMH

### LINKED ISSUE #79: Infinite reconciliation loop when waiting for provisioning
**What steps did you take and what happened:**
When a node is being provisioned, the controller keeps reconciling the metal3machine about every second. 

**What did you expect to happen:**

It should not reconcile the metal3machine until the BMH is updated, and probably provisioned.

**Anything else you would like to add:**
It looks like the call to updateObject function in https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/baremetal/metal3machine_manager.go#L651 updates the BMH object that in turn triggers a reconciliation of the metal3machine. Calls to the updateObject function for BMH should happen only when there has been a change on the BMH object. Some refactoring and the use of patch helper from CAPI would be very much needed to prevent this waste of resources.

**Environment:**

- Cluster-api version: 0.3.6
- CAPM3 version: 0.4.X
- environment (metal3-dev-env or other): metal3-dev-env

/kind bug

### PR DESCRIPTION
**Pull request by maelk**
**What this PR does / why we need it**:
Using the updateObject function was triggering an infinite loop,
as the metal3machine were reconciled when BMH were updated and
the reconciliation was always updating the BMH

**Which issue(s) this PR fixes** :
Fixes #79

### CODE REVIEW

File: baremetal/utils.go [hunk 1]
```diff
@@ -44,3 +53,26 @@ type NotFoundError struct {
...
 }
+
+func patchIfFound(ctx context.Context, helper *patch.Helper, host runtime.Object) error {
+	err := helper.Patch(ctx, host)
+	if err != nil {
+		notFound := true
+		if aggr, ok := err.(kerrors.Aggregate); ok {
+			for _, kerr := range aggr.Errors() {
+				if !apierrors.IsNotFound(kerr) {
+					notFound = false
```
Comment (kashifest) [inline baremetal/utils.go:64]: why not returning nil if `apierrors.IsNotFound(kerr)`

File: baremetal/metal3machine_manager.go [hunk 2]
```diff
@@ -576,8 +585,19 @@ func (m *MachineManager) exists(ctx context.Context) (bool, error) {
 // getHost gets the associated host by looking for an annotation on the machine
 // that contains a reference to the host. Returns nil if not found. Assumes the
 // host is in the same namespace as the machine.
-func (m *MachineManager) getHost(ctx context.Context) (*bmh.BareMetalHost, error) {
-	annotations := m.Metal3Machine.ObjectMeta.GetAnnotations()
+func (m *MachineManager) getHost(ctx context.Context) (*bmh.BareMetalHost, *patch.Helper, error) {
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:588]: a bit confused why do we need two functions with same name, if the idea is to get helper also, may be the naming could be changed accordingly?

File: baremetal/utils.go (hunk 1 above)
Comment (maelk) [reply baremetal/utils.go:64]: because it is already deleted, the end goal is achieved

File: baremetal/metal3machine_manager.go (hunk 2 above)
Comment (maelk) [reply baremetal/metal3machine_manager.go:588]: one of them takes the manager, the other is a helper, that can be used in other places. Having the same name is not a problem, and IMO it clearly shows the flow, you call the gethost of the manager, that under the hood calls the helper of the same name

File: baremetal/utils.go (hunk 1 above)
Comment (kashifest) [reply baremetal/utils.go:64]: not clear, you are anyhow returning nil when `if notFound`

Comment (maelk) [reply baremetal/utils.go:64]: oh, I see sorry, I did not get your question right. The problem is that you are getting an aggregate of errors, and I want to return nil if all errors are NotFound. If one of them is not NotFound, I want to return the normal error (or requeue). so we need to check them all, and notFound is true only if all of them are NotFound errors, in which case we return nil

**Pull request merged**